<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Hugo 0.55.5" />


<title>Limma Moderated and Ordinary t-statistics - Sahir&#39;s blog</title>
<meta property="og:title" content="Limma Moderated and Ordinary t-statistics - Sahir&#39;s blog">


  <link href='https://sahirbhatnagar.github.io/blog/favicon.ico' rel='icon' type='image/x-icon'/>



  








<link href='//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css' rel='stylesheet' type='text/css' />



<link rel="stylesheet" href="/blog/css/fonts.css" media="all">
<link rel="stylesheet" href="/blog/css/main.css" media="all">



  </head>
  <body>
    <div class="wrapper">
      <header class="header">
        <nav class="nav">
  <a href="/blog/" class="nav-logo">
    <img src="/blog/images/logo.png"
         width="50"
         height="50"
         alt="Logo">
  </a>

  <ul class="nav-links">
    
    <li><a href="https://sahirbhatnagar.github.io/">Home</a></li>
    
  </ul>
</nav>

      </header>


<main class="content" role="main">

  <article class="article">
    
    <span class="article-duration">3 min read</span>
    

    <h1 class="article-title">Limma Moderated and Ordinary t-statistics</h1>

    
    <span class="article-date">2017-02-07</span>
    

    <div class="article-content">
      <p>When analyzing large amounts of genetic and genomic data, the first line of analysis is usually some sort of univariate test. That is, conduct a statistical test for each SNP or CpG site or Gene and then correct for multiple testing. The <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> package on Bioconductor is a popular method for computing <em>moderated</em> t-statistics using a combination of the <code>limma::lmFit</code> and <code>limma::eBayes</code> functions. In this post, I show how to calculate the <em>ordinary</em> t-statistics from <code>limma</code> output.</p>

<p>First we load the required packages</p>

<pre class="r"><code>
# clear workspace
rm(list=ls())

if (!requireNamespace("pacman", quietly = TRUE)) 
  install.packages("pacman")

# this will install (if you dont already have it) and load the package
pacman::p_load("minfi")
pacman::p_load("minfiData")
pacman::p_load("limma")
pacman::p_load("CpGassoc")

# check the data that is available in the minfiData package
pacman::p_data("minfiData")
</code></pre>

<pre class="r"><code>
##   Data        Description                                                                           
## 1 MsetEx      An example dataset for Illumina's Human Methylation 450k dataset, after preprocessing.
## 2 MsetEx.sub  An example dataset for Illumina's Human Methylation 450k dataset, after preprocessing.
## 3 RGsetEx     An example dataset for Illumina's Human Methylation 450k dataset.                     
## 4 RGsetEx.sub An example dataset for Illumina's Human Methylation 450k dataset.
</code></pre>

<p>Next, we extract some sample data and create a covariate of interest</p>

<pre class="r"><code>
# get the M-values for the sample data
DT <- minfi::getM(MsetEx.sub)

dim(DT)
</code></pre>

<pre class="r"><code>
## [1] 600   6
</code></pre>

<pre class="r"><code>
# rows are CpGs and columns are samples
head(DT)
</code></pre>

<pre class="r"><code>
##            5723646052_R02C02 5723646052_R04C01 5723646052_R05C02
## cg00050873          3.502348         0.4414491          4.340695
## cg00212031         -3.273751         0.9234662         -2.614777
## cg00213748          2.076816        -0.1309465          1.260995
## cg00214611         -3.438838         1.7463950         -2.270551
## cg00455876          1.839010        -0.9927320          1.619479
## cg01707559         -3.303987        -0.6433201         -3.540887
##            5723646053_R04C02 5723646053_R05C02 5723646053_R06C02
## cg00050873        0.24458355        -0.3219281         0.2744392
## cg00212031       -0.21052257        -0.6861413        -0.1397595
## cg00213748       -1.10373279        -1.6616553        -0.1270869
## cg00214611        0.29029649        -0.2103599        -0.6138630
## cg00455876       -0.09504721        -0.2854655         0.6361273
## cg01707559       -0.74835377        -0.4678048        -1.1345421
</code></pre>

<pre class="r"><code>
# create a fake covariate. 3 intact and 3 degraded samples
tissue <- factor(c(rep("Intact",3), rep("Degraded",3)))
design <- model.matrix(~ tissue)
</code></pre>

<p>Then we calculate the moderated and ordinary t-statistics and compare them:</p>

<pre class="r"><code>
# limma fit
fit <- lmFit(DT, design)
fit_ebayes <- eBayes(fit)

# ordinary t-statistic
ordinary.t <- fit$coefficients[,"tissueIntact"] / 
  fit$stdev.unscaled[,"tissueIntact"] / 
  fit$sigma

plot(x = ordinary.t, 
     y = fit_ebayes$t[,"tissueIntact"], 
     xlab = "ordinary t-statistic", 
     ylab = "moderated t-statistic")
abline(a = 0, b = 1, col = "red")
</code></pre>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-2-1.png" alt="plot of chunk unnamed-chunk-2" /></p>

<p>We can calculate the corresponding p-values from the ordinary t-statistics. This is given by</p>

<pre class="r"><code>
# t-distribution with n-p-1 degrees of freedom 
# (n: number of samples, p:number of covariates excluding intercept)
ordinary.pvalue <- pt(q = abs(ordinary.t), 
                      df = dim(DT)[2]-(ncol(design)-1)-1, 
                      lower.tail = F) * 2
</code></pre>

<p>We can also use the <a href="https://cran.r-project.org/package=CpGassoc"><code>CpGassoc</code></a> package to calculate ordinary t-statistics and compare the result to our manual calculations:</p>

<pre class="r"><code>
# compare result with CpGassoc package (regular t-test)
t_test_mvalues <- CpGassoc::cpg.assoc(
  beta.val = minfi::getBeta(MsetEx.sub),
  indep = tissue,
  fdr.cutoff = 0.05,
  logit.transform = TRUE)

# these should be equal
plot(t_test_mvalues$results$P.value, ordinary.pvalue)
abline(a = 0, b = 1, col = "red")
</code></pre>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-4-1.png" alt="plot of chunk unnamed-chunk-4" /></p>

<pre class="r"><code>
# t-statistic is the squared F.statistic (in certain cases) 
plot(t_test_mvalues$results$F.statistic, ordinary.t^2)
abline(a = 0, b = 1, col = "red")
</code></pre>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-4-2.png" alt="plot of chunk unnamed-chunk-4" /></p>

<h2 id="session-info">Session Info</h2>

<pre class="r"><code>
## R version 3.3.2 (2016-10-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.1 LTS
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices
## [7] utils     datasets  base     
## 
## other attached packages:
##  [1] CpGassoc_2.55                                     
##  [2] nlme_3.1-128                                      
##  [3] limma_3.30.10                                     
##  [4] minfiData_0.20.0                                  
##  [5] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.0
##  [6] IlluminaHumanMethylation450kmanifest_0.4.0        
##  [7] minfi_1.20.2                                      
##  [8] bumphunter_1.14.0                                 
##  [9] locfit_1.5-9.1                                    
## [10] iterators_1.0.8                                   
## [11] foreach_1.4.3                                     
## [12] Biostrings_2.42.1                                 
## [13] XVector_0.14.0                                    
## [14] SummarizedExperiment_1.4.0                        
## [15] GenomicRanges_1.26.2                              
## [16] GenomeInfoDb_1.10.2                               
## [17] IRanges_2.8.1                                     
## [18] S4Vectors_0.12.1                                  
## [19] Biobase_2.34.0                                    
## [20] BiocGenerics_0.20.0                               
## [21] knitr_1.15.1                                      
## 
## loaded via a namespace (and not attached):
##  [1] mclust_5.2.2             base64_2.0              
##  [3] Rcpp_0.12.9              lattice_0.20-34         
##  [5] Rsamtools_1.26.1         digest_0.6.12           
##  [7] R6_2.2.0                 plyr_1.8.4              
##  [9] RSQLite_1.1-2            evaluate_0.10           
## [11] highr_0.6                httr_1.2.1              
## [13] zlibbioc_1.20.0          GenomicFeatures_1.26.2  
## [15] data.table_1.10.4        annotate_1.52.1         
## [17] Matrix_1.2-7.1           preprocessCore_1.36.0   
## [19] splines_3.3.2            BiocParallel_1.8.1      
## [21] stringr_1.2.0            RCurl_1.95-4.8          
## [23] biomaRt_2.30.0           rtracklayer_1.34.1      
## [25] multtest_2.30.0          pkgmaker_0.22           
## [27] openssl_0.9.6            GEOquery_2.40.0         
## [29] quadprog_1.5-5           codetools_0.2-15        
## [31] matrixStats_0.51.0       XML_3.98-1.5            
## [33] reshape_0.8.6            GenomicAlignments_1.10.0
## [35] MASS_7.3-45              bitops_1.0-6            
## [37] grid_3.3.2               xtable_1.8-2            
## [39] registry_0.3             DBI_0.5-1               
## [41] pacman_0.4.1             magrittr_1.5            
## [43] stringi_1.1.2            genefilter_1.56.0       
## [45] doRNG_1.6                nor1mix_1.2-2           
## [47] RColorBrewer_1.1-2       siggenes_1.48.0         
## [49] tools_3.3.2              illuminaio_0.16.0       
## [51] rngtools_1.2.4           survival_2.40-1         
## [53] AnnotationDbi_1.36.2     beanplot_1.2            
## [55] memoise_1.0.0
</code></pre>
    </div>
  </article>

  
<section id="comments">
  <div id="disqus_thread"></div>
  <script>
  var disqus_config = function () {
  
  };
  (function() {
    var inIFrame = function() {
      var iframe = true;
      try { iframe = window.self !== window.top; } catch (e) {}
      return iframe;
    };
    if (inIFrame()) return;
    var d = document, s = d.createElement('script');
    s.src = '//sahirbhatnagar.disqus.com/embed.js'; s.async = true;
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</section>



</main>

      <footer class="footer">
        <ul class="footer-links">
          <li>
            <a href="/blog/index.xml" type="application/rss+xml" target="_blank">RSS feed</a>
          </li>
          <li>
            <a href="https://gohugo.io/" class="footer-links-kudos">Made with <img src="/blog/images/hugo-logo.png" alt="Img link to Hugo website" width="22" height="22"></a>
          </li>
        </ul>
      </footer>

    </div>
    



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>



    
<script src="/blog/js/math-code.js"></script>
<script async src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>


    
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-60189402-1', 'auto');
	
	ga('send', 'pageview');
}
</script>

  </body>
</html>

