<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Hugo 0.54.0" />


<title>Limma Moderated and Ordinary t-statistics - Sahir&#39;s blog</title>
<meta property="og:title" content="Limma Moderated and Ordinary t-statistics - Sahir&#39;s blog">


  <link href='https://sahirbhatnagar.github.io/blog/favicon.ico' rel='icon' type='image/x-icon'/>



  







<link rel="stylesheet" href="/blog/css/fonts.css" media="all">
<link rel="stylesheet" href="/blog/css/main.css" media="all">



  </head>
  <body>
    <div class="wrapper">
      <header class="header">
        <nav class="nav">
  <a href="/blog/" class="nav-logo">
    <img src="/blog/images/logo.png"
         width="50"
         height="50"
         alt="Logo">
  </a>

  <ul class="nav-links">
    
    <li><a href="https://sahirbhatnagar.github.io/blog/">Home</a></li>
    
  </ul>
</nav>

      </header>


<main class="content" role="main">

  <article class="article">
    
    <span class="article-duration">4 min read</span>
    

    <h1 class="article-title">Limma Moderated and Ordinary t-statistics</h1>

    
    <span class="article-date">0001-01-01</span>
    

    <div class="article-content">
      <p>When analyzing large amounts of genetic and genomic data, the first line of analysis is usually some sort of univariate test. That is, conduct a statistical test for each SNP or CpG site or Gene and then correct for multiple testing. The <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> package on Bioconductor is a popular method for computing <em>moderated</em> t-statistics using a combination of the <code>limma::lmFit</code> and <code>limma::eBayes</code> functions. In this post, I show how to calculate the <em>ordinary</em> t-statistics from <code>limma</code> output.</p>

<p>First we load the required packages</p>

<p>{% highlight r %}</p>

<h1 id="clear-workspace">clear workspace</h1>

<p>rm(list=ls())</p>

<p>if (!requireNamespace(&ldquo;pacman&rdquo;, quietly = TRUE))
  install.packages(&ldquo;pacman&rdquo;)</p>

<h1 id="this-will-install-if-you-dont-already-have-it-and-load-the-package">this will install (if you dont already have it) and load the package</h1>

<p>pacman::p_load(&ldquo;minfi&rdquo;)
pacman::p_load(&ldquo;minfiData&rdquo;)
pacman::p_load(&ldquo;limma&rdquo;)
pacman::p_load(&ldquo;CpGassoc&rdquo;)</p>

<h1 id="check-the-data-that-is-available-in-the-minfidata-package">check the data that is available in the minfiData package</h1>

<p>pacman::p_data(&ldquo;minfiData&rdquo;)
{% endhighlight %}</p>

<p>{% highlight text %}</p>

<h2 id="data-description">Data        Description</h2>

<h2 id="1-msetex-an-example-dataset-for-illumina-s-human-methylation-450k-dataset-after-preprocessing">1 MsetEx      An example dataset for Illumina&rsquo;s Human Methylation 450k dataset, after preprocessing.</h2>

<h2 id="2-msetex-sub-an-example-dataset-for-illumina-s-human-methylation-450k-dataset-after-preprocessing">2 MsetEx.sub  An example dataset for Illumina&rsquo;s Human Methylation 450k dataset, after preprocessing.</h2>

<h2 id="3-rgsetex-an-example-dataset-for-illumina-s-human-methylation-450k-dataset">3 RGsetEx     An example dataset for Illumina&rsquo;s Human Methylation 450k dataset.</h2>

<h2 id="4-rgsetex-sub-an-example-dataset-for-illumina-s-human-methylation-450k-dataset">4 RGsetEx.sub An example dataset for Illumina&rsquo;s Human Methylation 450k dataset.</h2>

<p>{% endhighlight %}</p>

<p>Next, we extract some sample data and create a covariate of interest</p>

<p>{% highlight r %}</p>

<h1 id="get-the-m-values-for-the-sample-data">get the M-values for the sample data</h1>

<p>DT &lt;- minfi::getM(MsetEx.sub)</p>

<p>dim(DT)
{% endhighlight %}</p>

<p>{% highlight text %}</p>

<h2 id="1-600-6">[1] 600   6</h2>

<p>{% endhighlight %}</p>

<p>{% highlight r %}</p>

<h1 id="rows-are-cpgs-and-columns-are-samples">rows are CpGs and columns are samples</h1>

<p>head(DT)
{% endhighlight %}</p>

<p>{% highlight text %}</p>

<h2 id="5723646052-r02c02-5723646052-r04c01-5723646052-r05c02">5723646052_R02C02 5723646052_R04C01 5723646052_R05C02</h2>

<h2 id="cg00050873-3-502348-0-4414491-4-340695">cg00050873          3.502348         0.4414491          4.340695</h2>

<h2 id="cg00212031-3-273751-0-9234662-2-614777">cg00212031         -3.273751         0.9234662         -2.614777</h2>

<h2 id="cg00213748-2-076816-0-1309465-1-260995">cg00213748          2.076816        -0.1309465          1.260995</h2>

<h2 id="cg00214611-3-438838-1-7463950-2-270551">cg00214611         -3.438838         1.7463950         -2.270551</h2>

<h2 id="cg00455876-1-839010-0-9927320-1-619479">cg00455876          1.839010        -0.9927320          1.619479</h2>

<h2 id="cg01707559-3-303987-0-6433201-3-540887">cg01707559         -3.303987        -0.6433201         -3.540887</h2>

<h2 id="5723646053-r04c02-5723646053-r05c02-5723646053-r06c02">5723646053_R04C02 5723646053_R05C02 5723646053_R06C02</h2>

<h2 id="cg00050873-0-24458355-0-3219281-0-2744392">cg00050873        0.24458355        -0.3219281         0.2744392</h2>

<h2 id="cg00212031-0-21052257-0-6861413-0-1397595">cg00212031       -0.21052257        -0.6861413        -0.1397595</h2>

<h2 id="cg00213748-1-10373279-1-6616553-0-1270869">cg00213748       -1.10373279        -1.6616553        -0.1270869</h2>

<h2 id="cg00214611-0-29029649-0-2103599-0-6138630">cg00214611        0.29029649        -0.2103599        -0.6138630</h2>

<h2 id="cg00455876-0-09504721-0-2854655-0-6361273">cg00455876       -0.09504721        -0.2854655         0.6361273</h2>

<h2 id="cg01707559-0-74835377-0-4678048-1-1345421">cg01707559       -0.74835377        -0.4678048        -1.1345421</h2>

<p>{% endhighlight %}</p>

<p>{% highlight r %}</p>

<h1 id="create-a-fake-covariate-3-intact-and-3-degraded-samples">create a fake covariate. 3 intact and 3 degraded samples</h1>

<p>tissue &lt;- factor(c(rep(&ldquo;Intact&rdquo;,3), rep(&ldquo;Degraded&rdquo;,3)))
design &lt;- model.matrix(~ tissue)
{% endhighlight %}</p>

<p>Then we calculate the moderated and ordinary t-statistics and compare them:</p>

<p>{% highlight r %}</p>

<h1 id="limma-fit">limma fit</h1>

<p>fit &lt;- lmFit(DT, design)
fit_ebayes &lt;- eBayes(fit)</p>

<h1 id="ordinary-t-statistic">ordinary t-statistic</h1>

<p>ordinary.t &lt;- fit$coefficients[,&ldquo;tissueIntact&rdquo;] /
  fit$stdev.unscaled[,&ldquo;tissueIntact&rdquo;] /
  fit$sigma</p>

<p>plot(x = ordinary.t,
     y = fit_ebayes$t[,&ldquo;tissueIntact&rdquo;],
     xlab = &ldquo;ordinary t-statistic&rdquo;,
     ylab = &ldquo;moderated t-statistic&rdquo;)
abline(a = 0, b = 1, col = &ldquo;red&rdquo;)
{% endhighlight %}</p>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-2-1.png" alt="plot of chunk unnamed-chunk-2" /></p>

<p>We can calculate the corresponding p-values from the ordinary t-statistics. This is given by</p>

<p>{% highlight r %}</p>

<h1 id="t-distribution-with-n-p-1-degrees-of-freedom">t-distribution with n-p-1 degrees of freedom</h1>

<h1 id="n-number-of-samples-p-number-of-covariates-excluding-intercept">(n: number of samples, p:number of covariates excluding intercept)</h1>

<p>ordinary.pvalue &lt;- pt(q = abs(ordinary.t),
                      df = dim(DT)[2]-(ncol(design)-1)-1,
                      lower.tail = F) * 2
{% endhighlight %}</p>

<p>We can also use the <a href="https://cran.r-project.org/package=CpGassoc"><code>CpGassoc</code></a> package to calculate ordinary t-statistics and compare the result to our manual calculations:</p>

<p>{% highlight r %}</p>

<h1 id="compare-result-with-cpgassoc-package-regular-t-test">compare result with CpGassoc package (regular t-test)</h1>

<p>t_test_mvalues &lt;- CpGassoc::cpg.assoc(
  beta.val = minfi::getBeta(MsetEx.sub),
  indep = tissue,
  fdr.cutoff = 0.05,
  logit.transform = TRUE)</p>

<h1 id="these-should-be-equal">these should be equal</h1>

<p>plot(t_test_mvalues$results$P.value, ordinary.pvalue)
abline(a = 0, b = 1, col = &ldquo;red&rdquo;)
{% endhighlight %}</p>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-4-1.png" alt="plot of chunk unnamed-chunk-4" /></p>

<p>{% highlight r %}</p>

<h1 id="t-statistic-is-the-squared-f-statistic-in-certain-cases">t-statistic is the squared F.statistic (in certain cases)</h1>

<p>plot(t_test_mvalues$results$F.statistic, ordinary.t^2)
abline(a = 0, b = 1, col = &ldquo;red&rdquo;)
{% endhighlight %}</p>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-4-2.png" alt="plot of chunk unnamed-chunk-4" /></p>

<h2 id="session-info">Session Info</h2>

<p>{% highlight text %}</p>

<h2 id="r-version-3-3-2-2016-10-31">R version 3.3.2 (2016-10-31)</h2>

<h2 id="platform-x86-64-pc-linux-gnu-64-bit">Platform: x86_64-pc-linux-gnu (64-bit)</h2>

<h2 id="running-under-ubuntu-16-04-1-lts">Running under: Ubuntu 16.04.1 LTS</h2>

<h2 id="attached-base-packages">attached base packages:</h2>

<h2 id="1-stats4-parallel-methods-stats-graphics-grdevices">[1] stats4    parallel  methods   stats     graphics  grDevices</h2>

<h2 id="7-utils-datasets-base">[7] utils     datasets  base</h2>

<h2 id="other-attached-packages">other attached packages:</h2>

<h2 id="1-cpgassoc-2-55">[1] CpGassoc_2.55</h2>

<h2 id="2-nlme-3-1-128">[2] nlme_3.1-128</h2>

<h2 id="3-limma-3-30-10">[3] limma_3.30.10</h2>

<h2 id="4-minfidata-0-20-0">[4] minfiData_0.20.0</h2>

<h2 id="5-illuminahumanmethylation450kanno-ilmn12-hg19-0-6-0">[5] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.0</h2>

<h2 id="6-illuminahumanmethylation450kmanifest-0-4-0">[6] IlluminaHumanMethylation450kmanifest_0.4.0</h2>

<h2 id="7-minfi-1-20-2">[7] minfi_1.20.2</h2>

<h2 id="8-bumphunter-1-14-0">[8] bumphunter_1.14.0</h2>

<h2 id="9-locfit-1-5-9-1">[9] locfit_1.5-9.1</h2>

<h2 id="10-iterators-1-0-8">[10] iterators_1.0.8</h2>

<h2 id="11-foreach-1-4-3">[11] foreach_1.4.3</h2>

<h2 id="12-biostrings-2-42-1">[12] Biostrings_2.42.1</h2>

<h2 id="13-xvector-0-14-0">[13] XVector_0.14.0</h2>

<h2 id="14-summarizedexperiment-1-4-0">[14] SummarizedExperiment_1.4.0</h2>

<h2 id="15-genomicranges-1-26-2">[15] GenomicRanges_1.26.2</h2>

<h2 id="16-genomeinfodb-1-10-2">[16] GenomeInfoDb_1.10.2</h2>

<h2 id="17-iranges-2-8-1">[17] IRanges_2.8.1</h2>

<h2 id="18-s4vectors-0-12-1">[18] S4Vectors_0.12.1</h2>

<h2 id="19-biobase-2-34-0">[19] Biobase_2.34.0</h2>

<h2 id="20-biocgenerics-0-20-0">[20] BiocGenerics_0.20.0</h2>

<h2 id="21-knitr-1-15-1">[21] knitr_1.15.1</h2>

<h2 id="loaded-via-a-namespace-and-not-attached">loaded via a namespace (and not attached):</h2>

<h2 id="1-mclust-5-2-2-base64-2-0">[1] mclust_5.2.2             base64_2.0</h2>

<h2 id="3-rcpp-0-12-9-lattice-0-20-34">[3] Rcpp_0.12.9              lattice_0.20-34</h2>

<h2 id="5-rsamtools-1-26-1-digest-0-6-12">[5] Rsamtools_1.26.1         digest_0.6.12</h2>

<h2 id="7-r6-2-2-0-plyr-1-8-4">[7] R6_2.2.0                 plyr_1.8.4</h2>

<h2 id="9-rsqlite-1-1-2-evaluate-0-10">[9] RSQLite_1.1-2            evaluate_0.10</h2>

<h2 id="11-highr-0-6-httr-1-2-1">[11] highr_0.6                httr_1.2.1</h2>

<h2 id="13-zlibbioc-1-20-0-genomicfeatures-1-26-2">[13] zlibbioc_1.20.0          GenomicFeatures_1.26.2</h2>

<h2 id="15-data-table-1-10-4-annotate-1-52-1">[15] data.table_1.10.4        annotate_1.52.1</h2>

<h2 id="17-matrix-1-2-7-1-preprocesscore-1-36-0">[17] Matrix_1.2-7.1           preprocessCore_1.36.0</h2>

<h2 id="19-splines-3-3-2-biocparallel-1-8-1">[19] splines_3.3.2            BiocParallel_1.8.1</h2>

<h2 id="21-stringr-1-2-0-rcurl-1-95-4-8">[21] stringr_1.2.0            RCurl_1.95-4.8</h2>

<h2 id="23-biomart-2-30-0-rtracklayer-1-34-1">[23] biomaRt_2.30.0           rtracklayer_1.34.1</h2>

<h2 id="25-multtest-2-30-0-pkgmaker-0-22">[25] multtest_2.30.0          pkgmaker_0.22</h2>

<h2 id="27-openssl-0-9-6-geoquery-2-40-0">[27] openssl_0.9.6            GEOquery_2.40.0</h2>

<h2 id="29-quadprog-1-5-5-codetools-0-2-15">[29] quadprog_1.5-5           codetools_0.2-15</h2>

<h2 id="31-matrixstats-0-51-0-xml-3-98-1-5">[31] matrixStats_0.51.0       XML_3.98-1.5</h2>

<h2 id="33-reshape-0-8-6-genomicalignments-1-10-0">[33] reshape_0.8.6            GenomicAlignments_1.10.0</h2>

<h2 id="35-mass-7-3-45-bitops-1-0-6">[35] MASS_7.3-45              bitops_1.0-6</h2>

<h2 id="37-grid-3-3-2-xtable-1-8-2">[37] grid_3.3.2               xtable_1.8-2</h2>

<h2 id="39-registry-0-3-dbi-0-5-1">[39] registry_0.3             DBI_0.5-1</h2>

<h2 id="41-pacman-0-4-1-magrittr-1-5">[41] pacman_0.4.1             magrittr_1.5</h2>

<h2 id="43-stringi-1-1-2-genefilter-1-56-0">[43] stringi_1.1.2            genefilter_1.56.0</h2>

<h2 id="45-dorng-1-6-nor1mix-1-2-2">[45] doRNG_1.6                nor1mix_1.2-2</h2>

<h2 id="47-rcolorbrewer-1-1-2-siggenes-1-48-0">[47] RColorBrewer_1.1-2       siggenes_1.48.0</h2>

<h2 id="49-tools-3-3-2-illuminaio-0-16-0">[49] tools_3.3.2              illuminaio_0.16.0</h2>

<h2 id="51-rngtools-1-2-4-survival-2-40-1">[51] rngtools_1.2.4           survival_2.40-1</h2>

<h2 id="53-annotationdbi-1-36-2-beanplot-1-2">[53] AnnotationDbi_1.36.2     beanplot_1.2</h2>

<h2 id="55-memoise-1-0-0">[55] memoise_1.0.0</h2>

<p>{% endhighlight %}</p>
    </div>
  </article>

  
<section id="comments">
  <div id="disqus_thread"></div>
  <script>
  var disqus_config = function () {
  
  };
  (function() {
    var inIFrame = function() {
      var iframe = true;
      try { iframe = window.self !== window.top; } catch (e) {}
      return iframe;
    };
    if (inIFrame()) return;
    var d = document, s = d.createElement('script');
    s.src = '//sahirbhatnagar.disqus.com/embed.js'; s.async = true;
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</section>



</main>

      <footer class="footer">
        <ul class="footer-links">
          <li>
            <a href="/blog/index.xml" type="application/rss+xml" target="_blank">RSS feed</a>
          </li>
          <li>
            <a href="https://gohugo.io/" class="footer-links-kudos">Made with <img src="/blog/images/hugo-logo.png" alt="Img link to Hugo website" width="22" height="22"></a>
          </li>
        </ul>
      </footer>

    </div>
    

    
<script src="/blog/js/math-code.js"></script>
<script async src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>


    
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-60189402-1', 'auto');
	
	ga('send', 'pageview');
}
</script>

  </body>
</html>

